File: src/audio_steg_desktop/src/AudioSteg.Desktop/AppConfig.cs
using System.IO;
using System.Text.Json;

namespace AudioSteg.Desktop;

/// <summary>
/// Deploy-time flags from <c>appsettings.json</c> beside the executable (not user settings UI).
/// </summary>
public sealed class AppConfig
{
    /// <summary>When false, hides the upload-file button on the embed (steganography) tab.</summary>
    public bool ShowEmbedLoadFileButton { get; set; } = false;

    /// <summary>When true, shows recovery message-length dialog after embed.</summary>
    public bool ShowEmbedRecoveryDialog { get; set; } = true;

    /// <summary>Fixed msg_len when default message bit limit is enabled (2^18 = 262144).</summary>
    public int DefaultFixedMessageBitLength { get; set; } = 262144;

    /// <summary>Default logistic map r for new sessions / reset.</summary>
    public double LogisticR { get; set; } = 3.99;

    /// <summary>Default logistic map x0 for new sessions / reset.</summary>
    public double LogisticX0 { get; set; } = 0.45;

    public static AppConfig Current { get; private set; } = new();

    private static string AppSettingsFilePath =>
        Path.Combine(AppContext.BaseDirectory, "appsettings.json");

    public static void Load()
    {
        Current = new AppConfig();
        if (!File.Exists(AppSettingsFilePath)) return;
        try
        {
            var json = File.ReadAllText(AppSettingsFilePath);
            var loaded = JsonSerializer.Deserialize<AppConfig>(json);
            if (loaded is not null)
                Current = loaded;
        }
        catch
        {
            // Keep defaults when the file is invalid.
        }
    }
}

File: src/audio_stegano_app/lib/app/app_config.dart
import 'dart:convert';

import 'package:flutter/foundation.dart';
import 'package:flutter/services.dart';

import 'app_config_loader.dart';

/// Deploy-time configuration from [appsettings.json] beside the app or bundled asset.
class AppConfig {
  final bool showEmbedLoadFileButton;
  final bool showEmbedRecoveryDialog;

  /// Fixed msg_len when default message bit limit is enabled (2^18 = 262144).
  final int defaultFixedMessageBitLength;

  /// Default logistic map [r] for new sessions / reset.
  final double logisticR;

  /// Default logistic map [x0] for new sessions / reset.
  final double logisticX0;

  const AppConfig({
    required this.showEmbedLoadFileButton,
    required this.showEmbedRecoveryDialog,
    required this.defaultFixedMessageBitLength,
    required this.logisticR,
    required this.logisticX0,
  });

  /// UI flag: on Windows desktop, file load stays available when deploy flag is false.
  static bool showEmbedLoadFileForUiValue({
    required bool showEmbedLoadFileButton,
    required bool isNativeWindowsDesktop,
  }) =>
      showEmbedLoadFileButton || isNativeWindowsDesktop;

  bool get showEmbedLoadFileForUi => showEmbedLoadFileForUiValue(
        showEmbedLoadFileButton: showEmbedLoadFileButton,
        isNativeWindowsDesktop:
            !kIsWeb && defaultTargetPlatform == TargetPlatform.windows,
      );

  static const AppConfig defaults = AppConfig(
    showEmbedLoadFileButton: false,
    showEmbedRecoveryDialog: true,
    defaultFixedMessageBitLength: 262144,
    logisticR: 3.99,
    logisticX0: 0.45,
  );

  /// Asset keys used after [tryLoadDeployAppSettingsJson] (web / IO beside exe).
  static const List<String> bundledAssetPaths = [
    'assets/appsettings.json',
    '../../appsettings.json',
  ];

  static AppConfig _parseJson(String raw) {
    final map = jsonDecode(raw) as Map<String, dynamic>;
    return AppConfig(
      showEmbedLoadFileButton:
          map['ShowEmbedLoadFileButton'] as bool? ??
          defaults.showEmbedLoadFileButton,
      showEmbedRecoveryDialog:
          map['ShowEmbedRecoveryDialog'] as bool? ??
          defaults.showEmbedRecoveryDialog,
      defaultFixedMessageBitLength:
          (map['DefaultFixedMessageBitLength'] as num?)?.toInt() ??
          defaults.defaultFixedMessageBitLength,
      logisticR: (map['LogisticR'] as num?)?.toDouble() ?? defaults.logisticR,
      logisticX0:
          (map['LogisticX0'] as num?)?.toDouble() ?? defaults.logisticX0,
    );
  }

  static Future<String?> _loadBundledAsset() async {
    for (final path in bundledAssetPaths) {
      try {
        final raw = await rootBundle.loadString(path);
        if (raw.trim().isNotEmpty) {
          return raw;
        }
      } catch (_) {
        // Try next bundled path.
      }
    }
    return null;
  }

  static Future<AppConfig> load() async {
    try {
      final external = await tryLoadDeployAppSettingsJson();
      if (external != null && external.trim().isNotEmpty) {
        if (kDebugMode) {
          debugPrint('AppConfig: loaded from deploy file or web fetch');
        }
        return _parseJson(external);
      }

      final bundled = await _loadBundledAsset();
      if (bundled != null) {
        if (kDebugMode) {
          debugPrint('AppConfig: loaded from Flutter asset bundle');
        }
        return _parseJson(bundled);
      }
    } catch (e, st) {
      if (kDebugMode) {
        debugPrint('AppConfig.load failed: $e\n$st');
      }
    }

    if (kDebugMode) {
      debugPrint('AppConfig: using compiled defaults');
    }
    return defaults;
  }
}
